src/section/mod.rs

```diff
@@ -8,27 +8,395 @@
+
+        let next_bit = our_name.bit(next_bit_index);
+
+        let (our_new_size, sibling_new_size) = self
+            .members
+            .adults()
+            .map(|peer| peer.name().bit(next_bit_index) == next_bit)
+            .fold((0, 0), |(ours, siblings), is_our_prefix| {
+                if is_our_prefix {
+                    (ours + 1, siblings)
+                } else {
+                    (ours, siblings + 1)
+                }
+            });
+
+        // If either of the two new sections will not contain enough entries, return `false`.
+        if our_new_size < network_params.recommended_section_size
+            || sibling_new_size < network_params.recommended_section_size
+        {
+            return None;
+        }
+
+        let our_prefix = self.prefix().pushed(next_bit);
+        let other_prefix = self.prefix().pushed(!next_bit);
+
+        let our_elders = self.members.elder_candidates_matching_prefix(
+            &our_prefix,
+            network_params.elder_size,
+            self.elders_info(),
+        );
+        let other_elders = self.members.elder_candidates_matching_prefix(
+            &other_prefix,
+            network_params.elder_size,
+            self.elders_info(),
+        );
+
+        let our_info = EldersInfo::new(our_elders, our_prefix);
+        let other_info = EldersInfo::new(other_elders, other_prefix);
+
+        Some((our_info, other_info))
+    }
+
+    // Returns the candidates for elders out of all the nodes in the section, even out of the
+    // relocating nodes if there would not be enough instead.
+    fn elder_candidates(&self, elder_size: usize) -> BTreeMap<XorName, Peer> {
+        self.members
+            .elder_candidates(elder_size, self.elders_info())
+    }
+}
+
+// Create `EldersInfo` for the first node.
+fn create_first_elders_info(
+    pk_set: &bls::PublicKeySet,
+    sk_share: &bls::SecretKeyShare,
+    peer: Peer,
+) -> Result<Proven<EldersInfo>> {
+    let elders_info = EldersInfo::new(
+        iter::once((*peer.name(), peer)).collect(),
+        Prefix::default(),
+    );
+    let proof = create_first_proof(pk_set, sk_share, &elders_info)?;
+    Ok(Proven::new(elders_info, proof))
+}
+
+fn create_first_proof<T: Serialize>(
+    pk_set: &bls::PublicKeySet,
+    sk_share: &bls::SecretKeyShare,
+    payload: &T,
+) -> Result<Proof> {
+    let bytes = bincode::serialize(payload)?;
+    let signature_share = sk_share.sign(&bytes);
+    let signature = pk_set
+        .combine_signatures(iter::once((0, &signature_share)))
+        .map_err(|_| Error::InvalidSignatureShare)?;
+
+    Ok(Proof {
+        public_key: pk_set.public_key(),
+        signature,
+    })
+}
+
+fn cmp_section_chain_position<T: Serialize>(
+    lhs: &Proven<T>,
+    rhs: &Proven<T>,
+    section_chain: &SectionProofChain,
+) -> Option<Ordering> {
+    match (lhs.self_verify(), rhs.self_verify()) {
+        (true, true) => (),
+        (true, false) => return Some(Ordering::Greater),
```

Comment:
shall we still check the key of the valid proof is still in the chain ?

Reply:
We do. We call `index_of` a couple of lines below which returns `None` if the key is not present.